{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0f552cb0-1573-4caa-b372-8798d64789b0",
   "metadata": {},
   "source": [
    "## **PCR Primer Design using Python**\n",
    "**Written by Kilian Zindel**\n",
    "\n",
    "Polymerase Chain Reaction (PCR) is a method used to amplify (or make billions of copies of) a DNA strand. It is often a precursor to many other procedures used in genetic testing and research. In this method, DNA is replicated using repeated cycles of heating and cooling, called thermal cycles. There are three steps in each cycle:\n",
    "\n",
    "1. **Denaturation**: The temperature is raised to near boiling (~95°C), causing the double-stranded DNA to separate (or denature) into single strands.\n",
    "2. **Annealing**: The temperature is decreased to around 50–65°C, allowing **PCR primers** to bind (or anneal) to their complementary sequences on the DNA template.\n",
    "3. **Extension**: The temperature is raised to around 68–72°C, where a DNA polymerase enzyme binds to the **primers** and adds nucleotides to the 3' end, extending the DNA strand.\n",
    "\n",
    "This cycle is repeated multiple times, leading to exponential amplification of the target DNA. Under ideal conditions, 30 cycles can produce over a billion copies of the target DNA fragment.\n",
    "\n",
    "**PCR Primers** are short, single-stranded DNA sequences, typically spanning 18 to 25 nucleotides in length. They are essential for the extension phase of PCR because the Polymerase cannot start a new sequence, only add to the 3' end of an existing one. There are different kinds of primers designed for different use cases. In this tutorial I will be designing primers optimized for [Polymerase Chain Reaction (PCR)](https://www.biointeractive.org/classroom-resources/polymerase-chain-reaction-pcr). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "20bdb33d-a744-410c-9d7c-1ce6955ca031",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1073741824"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The power of exponential growth \n",
    "number_of_cycles = 30\n",
    "duplicate_DNA_strands = pow(2, number_of_cycles)\n",
    "duplicate_DNA_strands"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1d038b3-e5e0-4ba4-9c7b-ffe108975044",
   "metadata": {},
   "source": [
    "![PCR-image](https://upload.wikimedia.org/wikipedia/commons/thumb/a/ab/Polymerase_chain_reaction-en.svg/1024px-Polymerase_chain_reaction-en.svg.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ee040b1-b9d0-4a0e-982c-d489cc8ad777",
   "metadata": {},
   "source": [
    "#### Calculating the Melting Temperature (Tm) of a Primer\n",
    "\n",
    "An important consideration when designing a PCR primer is its melting temperature (Tm). During the annealing process, the temperature plays a critical role in how primers bind to the template DNA, as it affects the stability of hydrogen bonds between the primer and template. If the temperature is too high (e.g., >72°C), the hydrogen bonds break, and the primer begins to dissociate from the DNA. Conversely, if the temperature is too low (e.g., <50°C), primers are more likely to bind to and stick to non-target sequences.\n",
    "\n",
    "The temperature range of 50–65°C is considered optimal for the annealing temperature (Ta) in PCR. The exact annealing temperature should be set 3–5°C below the primer's melting temperature. This means the optimal melting temperature for the primer should be 55°C or higher.\n",
    "\n",
    "This temperature range is optimal for two main reasons:\n",
    "\n",
    "1. Specificity: The temperature is high enough for primers to anneal specifically to their complementary sequence on the template DNA.\n",
    "2. Stability: The temperature is low enough for primers to form stable bonds with the template DNA.\n",
    "\n",
    "The **melting temperature (Tm)** depends on three main factors:\n",
    "\n",
    "1. GC Content: Higher GC content leads to a higher melting temperature.\n",
    "2. Length: Longer primers have a higher melting temperature.\n",
    "3. Ionic Strength of the DNA Solution (total ion concentration): DNA is more stable in a solution with higher ionic strength.\n",
    "\n",
    "Studies have shown a linear relationship between primer GC content and melting temperature. The primer's melting temperature can be approximated using the **Wallace rule**. Below, we'll define two functions to calculate GC content and melting temperature.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3eac7809-b3ed-4458-8ea3-aa27861d6f33",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_gc_content(seq):\n",
    "    # calculate the percentage of nucleotides in sequence that are either 'G' or 'C'\n",
    "    c = seq.count('C')\n",
    "    g = seq.count('G')\n",
    "    gc_content = (g + c) / len(seq)\n",
    "    return round(100 * gc_content, 0) \n",
    "\n",
    "def calc_tm(seq):\n",
    "    # calculate the melting temperature of the sequence\n",
    "    # Simple approximation (Wallace rule):\n",
    "    # Tm ≈ 2°C*(A+T) + 4°C*(G+C)\n",
    "    a = seq.count('A')\n",
    "    t = seq.count('T')\n",
    "    g = seq.count('G')\n",
    "    c = seq.count('C')\n",
    "    return 2*(a+t) + 4*(g+c)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d5030ec-b150-4e98-9d92-d8a65b18c852",
   "metadata": {},
   "source": [
    "#### Detecting runs of repeated nucleotides\n",
    "We will need a function to detect runs of four or more identical bases. Avoiding such primers is considered best practice because excessive repeated bases can cause \"breathing\" of the primer, where a base bulges out. For example, a run of five guanine nucleotides might bind imperfectly to a run of four cytosine nucleotides, potentially leading to mispriming."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bd0cface-4edd-4eb7-b60f-46e234f7a65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_max_run(seq, max_run=3):\n",
    "    count = 1\n",
    "    for i in range(1, len(seq)):\n",
    "        # check if current and next base are the same\n",
    "        if seq[i-1] == seq[i]:\n",
    "            # if so: increment the count\n",
    "            count += 1\n",
    "            if count > max_run:\n",
    "                return True\n",
    "        else:\n",
    "            # otherwise: reset the count\n",
    "            count = 1 \n",
    "    return False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "734e6ae4-7bae-433a-9f15-9409b701e823",
   "metadata": {},
   "source": [
    "#### Checking the GC Clamp\n",
    "\n",
    "Because guanine (G) and cytosine (C) form three hydrogen bonds—compared to the two bonds formed by adenine (A) and thymine (T)—GC bonds provide greater stability for the primer-template bond. Having 1–3 G/C nucleotides at the 3' end of the primer helps to stabilize primer binding during PCR, reduces the likelihood of the primer dissociating prematurely, and helps facilitate extension by Taq polymerase.\n",
    "\n",
    "The 3' end of the primer, which I will define as the last five nucleotides, is especially important because this is where Taq polymerase begins synthesizing the new DNA strand. A stable 3' end improves the likilhood of a successful replication. \n",
    "\n",
    "However, we also need to ensure that we don't overclamping—have more than three G/C nucleotides in the last five bases—which can cause problems. Too much GC content at the 3' end increases the likelihood of primers forming secondary structures, for example: \n",
    "\n",
    "1. Primer dimers: Two primers bind to each other instead of the template DNA\n",
    "2. Hairpins: These are formed when the primer folds back on itself, creating a loop locked in by internal base pairing. \n",
    "\n",
    "In order to get an optimal primer, let's write a function which checks that the 3' end contains 1–3 G/C nucleotides and check whether the final nucleotide is a G or C."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ea340ab2-6238-4da4-95ce-a8360f4a301b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_gc_clamp(seq, max_gc_clamp=3):\n",
    "    # check the last nucleotide \n",
    "    last_nt = seq[-1]\n",
    "    if not (last_nt == 'G' or last_nt == 'C'):\n",
    "        return False\n",
    "    \n",
    "    # check the GC clamp \n",
    "    clamp = seq[-5:]\n",
    "    g = clamp.count('G')\n",
    "    c = clamp.count('C')\n",
    "    gc_content = g + c\n",
    "    if 1 <= gc_content <= max_gc_clamp:\n",
    "        # The GC Clamp is optimal \n",
    "        return True \n",
    "    else: \n",
    "        return False "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de9a905d-d550-489b-90b3-8a930ec2b48d",
   "metadata": {},
   "source": [
    "#### Best Practices in Selecting Primers\n",
    "- https://www.bocsci.com/resources/what-are-oligonucleotide-primers.html?srsltid=AfmBOoqrulSfsaXCtnkj0b8bwHnZRzN-KGALtljojbyG2wuYoq10FRNd\n",
    "- https://www.thermofisher.com/blog/behindthebench/pcr-primer-design-tips/\n",
    "\n",
    "There are a number of guidelines and best practices we should adhere to when selecting Primers\n",
    "- Primers typically range from 18 to 24 nucleotides in length\n",
    "- Optimal melting temperature ranges from 55°C to 65°C\n",
    "- Optimal GC content is between 40% and 60%\n",
    "- A stable 3' end which terminates in 2-3 G/C bases.\n",
    "- Avoid runs of more than 4 identical bases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7595a64-0b98-45f3-ba73-c18730a2737f",
   "metadata": {},
   "source": [
    "#### Now Let's go get some Sequences!\n",
    "Here are some websites with the info we need:\n",
    "- https://www.ncbi.nlm.nih.gov/gene\n",
    "- https://www.uniprot.org/uniprotkb/P25084/entry\n",
    "\n",
    "Instead of manually getting sequences on the webiste, we can use the python request library and the NCBI API to retrieve the information we need. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "35ddb456-0b9c-4241-9665-4b3d6d610eea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTTCATATGTTTTCGTCATAAATAAATAGTTACAAGAAATAATGGAGTCGTCTGACAATTTACATGATATAGATAATCTTGAAAACGGTAACATGGCTTGCCAGTGCTTCTTGGTTGGAGCCGGATACGTGGCTCTTGCAGCTGTGGCTTATCGTCTTTTGACGATTTTCTCGAATATTTTGGGCCCATACGTTCTTCTGTCGCCAATCGATTTGAAGAAAAGAGCTGGAGCTTCTTGGGCTGTTGTCACCGGAGCCACTGACGGAATCGGAAAAGCATACGCCTTCGAATTGGCTCGTCGTGGATTCAATGTCCTGCTCGTTTCGCGTACCCAATCAAAACTCGATGAGACGAAGAAGGAGATTCTCGAGAAGTATTCCAGCATTGAGGTCCGCACTGCCGCCTTCGACTTCACCAACGCTGCTCCTTCTGCTTACAAAGATCTTCTCGCCACCTTGAACCAAGTAGAGATCGGAGTTCTTATTAACAACGTTGGAATGAGCTACGAATATCCAGATGTACTTCACAAAGTTGACGGTGGAATCGAGCGTCTTGCAAACATCACCACCATCAACACTCTTCCACCAACATTGCTCTCCGCCGGAATCCTTCCACAAATGGTCGCACGAAAGGCTGGAGTCATTGTTAATGTTGGATCTTCAGCTGGAGCAAATCAAATGGCTCTCTGGGCTGTGTATTCAGCTACAAAGAAGTATGTCTCCTGGCTCACCGCTATCCTCCGAAAAGAATATGAACATCAAGGAATCACTGTCCAAACTATTGCTCCAATGATGGTCGCCACAAAGATGTCAAAAGTCAAGAGAACTTCATTCTTCACTCCAGACGGAGCCGTGTTCGCTAAATCAGCTCTGAACACTGTTGGAAATACCTCAGACACCACCGGATACATCACGCATCAACTTCAACTCGAGCTCATGGATCTCATTCCAACATTCATCCGCGACAAGATCCTCACAAATATGAGTGTCGGAACTCGTGCTGCTGCTCTCAGAAAGAAGGAAAGAGAAGCCAAATCTCAGTAATTTCTCGACGTTATTTTGTAGAATCTAAATCAGTCCAATCCATTCAATTATTGCTATTTATTTATGTTATATAAAGTTTTTGTTTTTACTAAAA'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def fetch_sequence(accession):\n",
    "    # gets FASTA data from the NCBI database using an accession number \n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
    "    params = {\n",
    "        'db': 'nucleotide',\n",
    "        'id': accession,\n",
    "        'rettype': 'fasta',\n",
    "        'retmode': 'text'\n",
    "    }\n",
    "    response = requests.get(base_url + \"efetch.fcgi\", params=params)\n",
    "    \n",
    "    # parse FASTA to extract sequence\n",
    "    fasta_data = response.text \n",
    "    # remove the first line and all newline characters\n",
    "    lines = fasta_data.splitlines()\n",
    "    sequence = \"\".join(lines[1:])\n",
    "    return sequence\n",
    "\n",
    "# Accession Number for Example Sequence: \n",
    "accession = \"NM_001268006.2\" # Caenorhabditis elegans act-1 gene,\n",
    "\n",
    "# Fetch Sequence from NCBI Database \n",
    "seq = fetch_sequence(accession)\n",
    "seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0df7dc28-c2ef-49c8-bd59-f6d55cf3ef70",
   "metadata": {},
   "source": [
    "#### Generating Candidates for Primers\n",
    "Let's set some parameters that we can use for our primer. We can define a ranges for the length, gc content, melting temperature etc. Then if our code doesn't generate any primers for the given sequence, we can go back and loosen the requirements. \n",
    "\n",
    "The following script uses a sliding window to iterate through every possible primer sequence of various lengths, then checks if each potential primer meets the constraints we've defined (optimal temperature, good clamp etc...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "698a8a7a-1c9a-4fec-b511-cf023f98f6d6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>GC%</th>\n",
       "      <th>Tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>19</td>\n",
       "      <td>TTGAGCTGGAACGCTCAAG</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>19</td>\n",
       "      <td>GAGCTGGAACGCTCAAGTG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>25</td>\n",
       "      <td>19</td>\n",
       "      <td>AGCTGGAACGCTCAAGTGG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start  length             sequence   GC%  Tm\n",
       "0     22      19  TTGAGCTGGAACGCTCAAG  53.0  58\n",
       "1     24      19  GAGCTGGAACGCTCAAGTG  58.0  60\n",
       "2     25      19  AGCTGGAACGCTCAAGTGG  58.0  60"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "sequence = \"ATGGCCTTGGTTGACGGTTTTCTTGAGCTGGAACGCTCAAGTGGAAAATTGGAGTGGAGCGCCATCCTGCAGAAGATGGCGAGCGACCTTGGATTCTCGAAGATCCTGTTCGGCCTGTTGCCTAAGGACAGCCAGGACTACGAGAACGCCTTCATCGTCGGCAACTACCCGGCCGCCTGGCGCGAGCATTACGACCGGGCTGGCTACGCGCGGGTCGACCCGACGGTCAGTCACTGTACCCAGAGCGTACTGCCGATTTTCTGGGAACCGTCCATCTACCAGACGCGAAAGCAGCACGAGTTCTTCGAGGAAGCCTCGGCCGCCGGCCTGGTGTATGGGCTGACCATGCCGCTGCATGGTGCTCGCGGCGAACTCGGCGCGCTGAGCCTCAGCGTGGAAGCGGAAAACCGGGCCGAGGCCAACCGTTTCATGGAGTCGGTCCTGCCGACCCTGTGGATGCTCAAGGACTACGCACTGCAGAGCGGTGCCGGACTGGCCTTCGAACATCCGGTCAGCAAACCGGTGGTTCTGACCAGCCGGGAGAAGGAAGTGTTGCAGTGGTGCGCCATCGGCAAGACCAGTTGGGAGATATCGGTTATCTGCAACTGCTCGGAAGCCAATGTGAACTTCCATATGGGAAATATTCGGCGGAAGTTCGGTGTGACCTCCCGCCGCGTAGCGGCCATTATGGCCGTTAATTTGGGTCTTATTACTCTCTGA\"\n",
    "# Set Paramaters\n",
    "min_length = 18\n",
    "max_length = 24\n",
    "gc_range = (40, 60)  # GC content range in %\n",
    "tm_range = (57, 62)  # Melting temperature range in °C\n",
    "max_self_complementarity = 5  # Maximum allowed score for self-complementarity\n",
    "max_gc_clamp = 3  # Max GC bases at the 3' end to avoid \"GC clamp\"\n",
    "max_hairpin_delta_g = -9.0  # Threshold for hairpin stability (kcal/mol)\n",
    "target = (50,250) # target \n",
    "\n",
    "# Generate Primer Candidates \n",
    "max_length = 24\n",
    "candidates = [] \n",
    "for length in range(min_length, max_length+1):\n",
    "    for start in range(target[0]-length):\n",
    "        primer = sequence[start:start+length]\n",
    "        \n",
    "        # check primer to ensure it meets constraints\n",
    "        if not check_gc_clamp(primer):\n",
    "            continue \n",
    "        \n",
    "        if check_max_run(primer):\n",
    "            continue \n",
    "\n",
    "        gc = calc_gc_content(primer)\n",
    "        if not (gc > gc_range[0] and gc < gc_range[1]):\n",
    "            continue \n",
    "        \n",
    "        tm = calc_tm(primer)\n",
    "        if not (tm > tm_range[0] and tm < tm_range[1]):\n",
    "            continue\n",
    "\n",
    "        candidates.append({\n",
    "            'start': start,\n",
    "            'length': length,\n",
    "            'sequence': primer,\n",
    "            'GC%': gc,\n",
    "            'Tm': tm,\n",
    "        })\n",
    "\n",
    "df = pd.DataFrame(candidates)\n",
    "df      "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c425e35e-6853-4401-86c6-e82db8c3fc84",
   "metadata": {},
   "source": [
    "#### Forward and Reverse Primers\n",
    "\n",
    "To amplify a DNA sequence in PCR, two primers are needed. A forward primer, to bind to the template strand, and a reverse primer, to bind to the complementary strand. The two primers should flank the target sequence on either side, with the 3' ends facing eachother.\n",
    "\n",
    "![primer_and_target_image](https://pressbooks.bccampus.ca/kathleef/wp-content/uploads/sites/1093/2020/08/Primers-on-target-DNA-png-4-1024x309.png)\n",
    "Designing the reverse primer requires an extra step. The reverse primer should match the sequence complementary to the one we are using to design it, so we will need to take the reverse complement of the primer when we test it. Let's write a function to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "737a906a-c02b-4f12-b5f8-29eb0a3da1f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(seq):    \n",
    "    complement_map = {'A':'T','T':'A','C':'G','G':'C','N':'N'}\n",
    "    rc = ''.join(complement_map.get(base, 'N') for base in reversed(seq))\n",
    "    return rc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06c84ca3-a577-49b0-9911-e4a907e6bb3e",
   "metadata": {},
   "source": [
    "To ensure the forward primer is positioned as close as possible to the beginning of the target sequence, the script starts at the beginning of the sequence and iterates backwards in the 5' direction. For the reverse primer, the script starts at the end of the target region and moves forward in the 3' direction, but once a primer is identified, its reverse complement is calculated and evaluated against the design constraints.\n",
    "\n",
    "The script is similar to the previous one but incorporates these directional sliding windows. It first identifies forward primers, storing the best candidates in a list. Then it searches for reverse primers, repeating the same process.\n",
    "\n",
    "A sliding window is used to iterate through every possible primer surrounding the target region. Each potential primer is checked to ensure it complies with our constraints, such as GC content, melting temperature, etc. Valid primers are stored in a data struture."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8bec9a4-a405-4fe0-9266-89bf69c82640",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>type</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>GC%</th>\n",
       "      <th>Tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>forward</td>\n",
       "      <td>19</td>\n",
       "      <td>TTGAGCTGGAACGCTCAAG</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>264</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>TCTGGTAGATGGACGGTTC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>271</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>TTTCGCGTCTGGTAGATGG</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>272</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>CTTTCGCGTCTGGTAGATG</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>285</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>AGAACTCGTGCTGCTTTCG</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>294</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>CTTCCTCGAAGAACTCGTG</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>297</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>AGGCTTCCTCGAAGAACTC</td>\n",
       "      <td>53.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>262</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>TGGTAGATGGACGGTTCCC</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>269</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>TCGCGTCTGGTAGATGGAC</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>263</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>CTGGTAGATGGACGGTTCC</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>forward</td>\n",
       "      <td>19</td>\n",
       "      <td>GAGCTGGAACGCTCAAGTG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>275</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>CTGCTTTCGCGTCTGGTAG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>284</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>GAACTCGTGCTGCTTTCGC</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>285</td>\n",
       "      <td>reverse</td>\n",
       "      <td>20</td>\n",
       "      <td>AAGAACTCGTGCTGCTTTCG</td>\n",
       "      <td>50.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>286</td>\n",
       "      <td>reverse</td>\n",
       "      <td>20</td>\n",
       "      <td>GAAGAACTCGTGCTGCTTTC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>25</td>\n",
       "      <td>forward</td>\n",
       "      <td>19</td>\n",
       "      <td>AGCTGGAACGCTCAAGTGG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>291</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>CCTCGAAGAACTCGTGCTG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>296</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>GGCTTCCTCGAAGAACTCG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>299</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>CGAGGCTTCCTCGAAGAAC</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>327</td>\n",
       "      <td>reverse</td>\n",
       "      <td>19</td>\n",
       "      <td>TGGTCAGCCCATACACCAG</td>\n",
       "      <td>58.0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    start     type  length              sequence   GC%  Tm\n",
       "2      22  forward      19   TTGAGCTGGAACGCTCAAG  53.0  58\n",
       "5     264  reverse      19   TCTGGTAGATGGACGGTTC  53.0  58\n",
       "7     271  reverse      19   TTTCGCGTCTGGTAGATGG  53.0  58\n",
       "8     272  reverse      19   CTTTCGCGTCTGGTAGATG  53.0  58\n",
       "11    285  reverse      19   AGAACTCGTGCTGCTTTCG  53.0  58\n",
       "15    294  reverse      19   CTTCCTCGAAGAACTCGTG  53.0  58\n",
       "17    297  reverse      19   AGGCTTCCTCGAAGAACTC  53.0  58\n",
       "3     262  reverse      19   TGGTAGATGGACGGTTCCC  58.0  60\n",
       "6     269  reverse      19   TCGCGTCTGGTAGATGGAC  58.0  60\n",
       "4     263  reverse      19   CTGGTAGATGGACGGTTCC  58.0  60\n",
       "1      24  forward      19   GAGCTGGAACGCTCAAGTG  58.0  60\n",
       "9     275  reverse      19   CTGCTTTCGCGTCTGGTAG  58.0  60\n",
       "10    284  reverse      19   GAACTCGTGCTGCTTTCGC  58.0  60\n",
       "12    285  reverse      20  AAGAACTCGTGCTGCTTTCG  50.0  60\n",
       "13    286  reverse      20  GAAGAACTCGTGCTGCTTTC  50.0  60\n",
       "0      25  forward      19   AGCTGGAACGCTCAAGTGG  58.0  60\n",
       "14    291  reverse      19   CCTCGAAGAACTCGTGCTG  58.0  60\n",
       "16    296  reverse      19   GGCTTCCTCGAAGAACTCG  58.0  60\n",
       "18    299  reverse      19   CGAGGCTTCCTCGAAGAAC  58.0  60\n",
       "19    327  reverse      19   TGGTCAGCCCATACACCAG  58.0  60"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence = \"ATGGCCTTGGTTGACGGTTTTCTTGAGCTGGAACGCTCAAGTGGAAAATTGGAGTGGAGCGCCATCCTGCAGAAGATGGCGAGCGACCTTGGATTCTCGAAGATCCTGTTCGGCCTGTTGCCTAAGGACAGCCAGGACTACGAGAACGCCTTCATCGTCGGCAACTACCCGGCCGCCTGGCGCGAGCATTACGACCGGGCTGGCTACGCGCGGGTCGACCCGACGGTCAGTCACTGTACCCAGAGCGTACTGCCGATTTTCTGGGAACCGTCCATCTACCAGACGCGAAAGCAGCACGAGTTCTTCGAGGAAGCCTCGGCCGCCGGCCTGGTGTATGGGCTGACCATGCCGCTGCATGGTGCTCGCGGCGAACTCGGCGCGCTGAGCCTCAGCGTGGAAGCGGAAAACCGGGCCGAGGCCAACCGTTTCATGGAGTCGGTCCTGCCGACCCTGTGGATGCTCAAGGACTACGCACTGCAGAGCGGTGCCGGACTGGCCTTCGAACATCCGGTCAGCAAACCGGTGGTTCTGACCAGCCGGGAGAAGGAAGTGTTGCAGTGGTGCGCCATCGGCAAGACCAGTTGGGAGATATCGGTTATCTGCAACTGCTCGGAAGCCAATGTGAACTTCCATATGGGAAATATTCGGCGGAAGTTCGGTGTGACCTCCCGCCGCGTAGCGGCCATTATGGCCGTTAATTTGGGTCTTATTACTCTCTGA\"\n",
    "min_length=18\n",
    "max_length=24\n",
    "gc_range = (40, 60)\n",
    "tm_range = (57, 62)\n",
    "max_self_complementary = 5\n",
    "max_gc_clamp = 3\n",
    "target = (50,250)\n",
    "\n",
    "primer_candidates = [] \n",
    "nt = target[0]\n",
    "\n",
    "while (len(primer_candidates) < 10 and nt > 0):\n",
    "    for length in range(min_length, max_length):\n",
    "        primer_start = nt - length - 1\n",
    "        if primer_start < 0:\n",
    "            continue \n",
    "        primer = sequence[primer_start:primer_start + length]\n",
    "        \n",
    "        # check primer to ensure it meets constraints\n",
    "        if not check_gc_clamp(primer):\n",
    "            continue \n",
    "        if check_max_run(primer):\n",
    "            continue \n",
    "        gc = calc_gc_content(primer)\n",
    "        if not (gc > gc_range[0] and gc < gc_range[1]):\n",
    "            continue     \n",
    "        tm = calc_tm(primer)\n",
    "        if not (tm > tm_range[0] and tm < tm_range[1]):\n",
    "            continue\n",
    "        primer_candidates.append({\n",
    "            'start': primer_start,\n",
    "            'type': 'forward',\n",
    "            'length': length,\n",
    "            'sequence': primer,\n",
    "            'GC%': gc,\n",
    "            'Tm': tm,\n",
    "        })\n",
    "    nt -= 1\n",
    "\n",
    "reverse_primer_candidates = [] \n",
    "nt = target[1]\n",
    "\n",
    "while (len(primer_candidates) < 20 and nt < len(sequence)):\n",
    "    for length in range(min_length, max_length):\n",
    "        primer_start = nt\n",
    "        if primer_start > len(sequence) - length - 1:\n",
    "            continue \n",
    "        primer = sequence[primer_start:primer_start + length]\n",
    "        primer = reverse_complement(primer) \n",
    "        # check primer to ensure it meets constraints\n",
    "        if not check_gc_clamp(primer):\n",
    "            continue \n",
    "        if check_max_run(primer):\n",
    "            continue \n",
    "        gc = calc_gc_content(primer)\n",
    "        if not (gc > gc_range[0] and gc < gc_range[1]):\n",
    "            continue     \n",
    "        tm = calc_tm(primer)\n",
    "        if not (tm > tm_range[0] and tm < tm_range[1]):\n",
    "            continue\n",
    "        primer_candidates.append({\n",
    "            'start': primer_start,\n",
    "            'type': 'reverse',\n",
    "            'length': length,\n",
    "            'sequence': primer,\n",
    "            'GC%': gc,\n",
    "            'Tm': tm,\n",
    "        })\n",
    "    nt += 1\n",
    "\n",
    "df = pd.DataFrame(primer_candidates)\n",
    "df = df.sort_values(by='Tm') # sort by termpature\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60e57eb9-4693-4930-bcb9-eeeafcc4e883",
   "metadata": {},
   "source": [
    "#### Choosing our Primer\n",
    "\n",
    "From the list of candidates, we can now select our forward and reverse primers. The melting temperatures (Tm) of the forward and reverse primers should be similar, ideally within 2–3°C of each other. To make this easier, the list has been sorted by melting temperature. Based on this criterion, Primer 1 and Primer 4 seem like strong candidates, they have the same melting temperature and both tightly flank the target region:\n",
    "\n",
    "- **Possible Forward Primer: GAGCTGGAACGCTCAAGTG**\n",
    "- **Possible Reverse Primer: CTGGTAGATGGACGGTTCC**\n",
    "\n",
    "While this program generates working primers, there are a couple ways it could be improved. A robust primer design tool should also check for secondary structures, such as:\n",
    "\n",
    "1. Hairpins: Structures formed when primers fold back on themselves.\n",
    "2. Primer Dimers: Primers bind to each other instead of the template DNA.\n",
    "\n",
    "Detecting these issues requires more sophisticated algorithms. Most Primer Design tools use complex thermodynamic models to more accurately calculate melting temperature (Tm) and predict secondary structures. These models account for interactions between bases and other factors affecting primer stability.\n",
    "\n",
    "Finally, the last step in primer design is ensuring the primer is specific. This requires checking that primers do not bind to unintended other of the genome, which could happen if the exact sequence (or a similar one) is repeated elsewhere. Specificity checks are performed using a BLAST (Basic Local Alignment Search Tool) search against the target genome. This step helps to avoid off-target amplification and ensures the primers are unique to the desired region. The NIH website has a free to use BLAST tool that can search their database.\n",
    "\n",
    "- https://blast.ncbi.nlm.nih.gov/Blast.cgi\n",
    "\n",
    "Overal, this project implements the basic principles of primer design for PCR using Python, providing a script that can generate and evaluate potential primers based on criteria such as desired melting temperature, gc content etc. This project introduced me to the complexities of primer design and helped me better understand the value of computation in biology. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3538875-ebe5-42ac-9f79-4f44c52ec33f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
